Failed v2.0.0 compilation on Linux arm64

When installing `cyksuid` in an arm64 (aarch64) Linux docker image, the build fails:

```
DEBUG:docker_builder:#21 46.63   Building wheel for cyksuid (pyproject.toml): started
DEBUG:docker_builder:#21 46.80   Building wheel for cyksuid (pyproject.toml): finished with status 'error'
DEBUG:docker_builder:#21 46.81   error: subprocess-exited-with-error
DEBUG:docker_builder:#21 46.81
DEBUG:docker_builder:#21 46.81   × Building wheel for cyksuid (pyproject.toml) did not run successfully.
DEBUG:docker_builder:#21 46.81   │ exit code: 1
DEBUG:docker_builder:#21 46.81   ╰─> [97 lines of output]
DEBUG:docker_builder:#21 46.81       running bdist_wheel
DEBUG:docker_builder:#21 46.81       running build
DEBUG:docker_builder:#21 46.81       running build_py
DEBUG:docker_builder:#21 46.81       creating build
DEBUG:docker_builder:#21 46.81       creating build/lib.linux-aarch64-cpython-39
DEBUG:docker_builder:#21 46.81       creating build/lib.linux-aarch64-cpython-39/cyksuid
DEBUG:docker_builder:#21 46.81       copying cyksuid/ksuid.py -> build/lib.linux-aarch64-cpython-39/cyksuid
DEBUG:docker_builder:#21 46.81       copying cyksuid/__init__.py -> build/lib.linux-aarch64-cpython-39/cyksuid
DEBUG:docker_builder:#21 46.81       copying cyksuid/__version__.py -> build/lib.linux-aarch64-cpython-39/cyksuid
DEBUG:docker_builder:#21 46.81       copying cyksuid/hints.py -> build/lib.linux-aarch64-cpython-39/cyksuid
DEBUG:docker_builder:#21 46.81       copying cyksuid/v2.py -> build/lib.linux-aarch64-cpython-39/cyksuid
DEBUG:docker_builder:#21 46.81       copying cyksuid/fast_base62.pyx -> build/lib.linux-aarch64-cpython-39/cyksuid
DEBUG:docker_builder:#21 46.81       copying cyksuid/_ksuid.pyx -> build/lib.linux-aarch64-cpython-39/cyksuid
DEBUG:docker_builder:#21 46.81       copying cyksuid/fast_base62.pxd -> build/lib.linux-aarch64-cpython-39/cyksuid
DEBUG:docker_builder:#21 46.81       copying cyksuid/_ksuid.pxd -> build/lib.linux-aarch64-cpython-39/cyksuid
DEBUG:docker_builder:#21 46.81       copying cyksuid/_ksuid.pyi -> build/lib.linux-aarch64-cpython-39/cyksuid
DEBUG:docker_builder:#21 46.81       copying cyksuid/fast_base62.pyi -> build/lib.linux-aarch64-cpython-39/cyksuid
DEBUG:docker_builder:#21 46.81       copying cyksuid/py.typed -> build/lib.linux-aarch64-cpython-39/cyksuid
DEBUG:docker_builder:#21 46.81       copying cyksuid/ksuidlite.h -> build/lib.linux-aarch64-cpython-39/cyksuid
DEBUG:docker_builder:#21 46.81       copying cyksuid/cbase62.h -> build/lib.linux-aarch64-cpython-39/cyksuid
DEBUG:docker_builder:#21 46.81       copying cyksuid/cbase62.cc -> build/lib.linux-aarch64-cpython-39/cyksuid
DEBUG:docker_builder:#21 46.81       copying cyksuid/fast_base62.cpp -> build/lib.linux-aarch64-cpython-39/cyksuid
DEBUG:docker_builder:#21 46.81       copying cyksuid/_ksuid.cpp -> build/lib.linux-aarch64-cpython-39/cyksuid
DEBUG:docker_builder:#21 46.81       running build_ext
DEBUG:docker_builder:#21 46.81       building 'cyksuid.fast_base62' extension
DEBUG:docker_builder:#21 46.81       creating build/temp.linux-aarch64-cpython-39
DEBUG:docker_builder:#21 46.81       creating build/temp.linux-aarch64-cpython-39/cyksuid
DEBUG:docker_builder:#21 46.81       gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -I/usr/local/include/python3.9 -c cyksuid/cbase62.cc -o build/temp.linux-aarch64-cpython-39/cyksuid/cbase62.o -Wno-write-strings -Wno-invalid-offsetof -Wno-sign-compare -Wno-unused-variable -std=c++14
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81          27 | };
DEBUG:docker_builder:#21 46.81             | ^
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       cyksuid/cbase62.cc:27:1: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
DEBUG:docker_builder:#21 46.81       error: command '/usr/bin/gcc' failed with exit code 1
DEBUG:docker_builder:#21 46.81       [end of output]
DEBUG:docker_builder:#21 46.81
DEBUG:docker_builder:#21 46.81   note: This error originates from a subprocess, and is likely not a problem with pip.
DEBUG:docker_builder:#21 46.81   ERROR: Failed building wheel for cyksuid
```

As a side note, it appears that the build is not generating wheels for Linux arm64.

--- a/.github/workflows/wheels.yml
+++ b/.github/workflows/wheels.yml
@@ -126,10 +126,22 @@ jobs:
               arch: x86_64
               build: "*manylinux*"
 
+          # - os: ubuntu-20.04
+          #   name: manylinux-aarch64
+          #   cibw:
+          #     arch: aarch64
+          #     build: "*manylinux*"
+
           - os: ubuntu-20.04
             name: musllinux
             cibw:
               build: "*musllinux*"
+
+          # - os: ubuntu-20.04
+          #   name: musllinux-arrch64
+          #   cibw:
+          #     arch: aarch64
+          #     build: "*musllinux*"
 
           - os: windows-2019
             name: win32
@@ -151,6 +163,13 @@ jobs:
 
     steps:
       - uses: actions/checkout@v3
+
+      # - name: Set up qemu
+      #   id: qemu
+      #   if: contains(matrix.os, 'ubuntu-') && matrix.cibw.arch == 'aarch64'
+      #   uses: docker/setup-qemu-action@v1
+      #   with:
+      #     platforms: arm64
 
       - name: setup python
         uses: actions/setup-python@v4

--- a/.gitignore
+++ b/.gitignore
@@ -18,3 +18,4 @@ bench.json
 
 .coverage
 coverage.xml
+a.out

--- a/cyksuid/_ksuid.pyx
+++ b/cyksuid/_ksuid.pyx
@@ -68,8 +68,7 @@ cdef class _KsuidMixin(object):
     @property
     def datetime(self):
         """Timestamp portion of the ID as a datetime.datetime object."""
-        cdef double ts = self.timestamp
-        return datetime.fromtimestamp(ts, tz=timezone.utc)
+        return datetime.fromtimestamp(self.timestamp, tz=timezone.utc)
 
     @property
     def timestamp_millis(self):

--- a/cyksuid/cbase62.cc
+++ b/cyksuid/cbase62.cc
@@ -13,16 +13,16 @@
 
 static constexpr char table_b2a_base62[] =
     "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
-static constexpr char table_a2b_base62[] = {
+static constexpr unsigned char table_a2b_base62[] = {
     /* clang-format off */
-    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
-    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
-    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
-    0, 1, 2, 3, 4, 5, 6, 7, 8, 9, -1, -1, -1, -1, -1, -1,
-    -1, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24,
-    25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, -1, -1, -1, -1, -1,
-    -1, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
-    51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, -1, -1, -1, -1, -1,
+    0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+    0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+    0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+    0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+    0xff, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24,
+    25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 0xff, 0xff, 0xff, 0xff, 0xff,
+    0xff, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
+    51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 0xff, 0xff, 0xff, 0xff, 0xff,
     /* clang-format on */
 };
 
@@ -96,7 +96,7 @@ int ksuid_b62_decode(unsigned char* dst, size_t dst_size, const char* src, size_
   for (size_t i = 0; i < _BASE62_ENCODED_SIZE; i++) {
     uint8_t c = src[i];
     uint8_t v = table_a2b_base62[c & 0x7f];
-    if ((c >= 0x80) || (v == (unsigned char)-1)) {
+    if (c >= 0x80) {
       return ERR_B62_INVALID_INPUT;
     }
 

--- a/setup.py
+++ b/setup.py
@@ -1,24 +1,23 @@
 import os
 import os.path
 import re
+import subprocess
 import sys
 import sysconfig
+import warnings
 from distutils import util
-from typing import Any, Dict, List, Tuple
+from distutils.command import build_ext
+from typing import Any, Dict, List, Optional, Tuple, Union, cast
 
 import pkg_resources
 from setuptools import Extension, setup
 
-IS_DEBUG = os.getenv("CYKSUID_DEBUG", "")
+IS_DEBUG = os.getenv("CYKSUID_DEBUG", "") == "1"
 
 ext_macros: List[Tuple[str, str]] = []
-ext_cythonize_kwargs: Dict[str, Any] = {
-    "force": True,
-    "compiler_directives": {},
-}
 ext_include_dirs = []
 
-if IS_DEBUG == "1":
+if IS_DEBUG:
     print("Build in debug mode")
     # Adding cython line trace for coverage report
     ext_macros += ("CYTHON_TRACE_NOGIL", "1"), ("CYTHON_TRACE", "1")
@@ -26,17 +25,6 @@ if IS_DEBUG == "1":
     # tests inside the cython package
     ext_include_dirs += [".."]
     # Some extra info for cython compiler
-    ext_cythonize_kwargs.update(
-        {
-            "gdb_debug": True,
-            "annotate": True,
-            "compiler_directives": {
-                "linetrace": True,
-                "profile": True,
-                "binding": True,
-            },
-        }
-    )
 
 
 try:
@@ -75,38 +63,6 @@ if USE_CYTHON:
 else:
     suffix = ".cpp"
 
-extra_compile_args: List[str] = []
-if sys.platform != "win32":
-    extra_compile_args.append("-Wno-write-strings")
-    extra_compile_args.append("-Wno-invalid-offsetof")
-    extra_compile_args.append("-Wno-sign-compare")
-    extra_compile_args.append("-Wno-unused-variable")
-    extra_compile_args.append("-std=c++14")
-
-if sys.platform == "darwin":
-    extra_compile_args.append("-Wno-shorten-64-to-32")
-    extra_compile_args.append("-Wno-deprecated-register")
-
-    # https://developer.apple.com/documentation/xcode_release_notes/xcode_10_release_notes
-    # C++ projects must now migrate to libc++ and are recommended to set a
-    # deployment target of macOS 10.9 or later, or iOS 7 or later.
-    if sys.platform == "darwin":
-        mac_target = str(sysconfig.get_config_var("MACOSX_DEPLOYMENT_TARGET"))
-        if mac_target and (
-            pkg_resources.parse_version(mac_target)
-            < pkg_resources.parse_version("10.9.0")
-        ):
-            os.environ["MACOSX_DEPLOYMENT_TARGET"] = "10.9"
-            os.environ["_PYTHON_HOST_PLATFORM"] = re.sub(
-                r"macosx-[0-9]+\.[0-9]+-(.+)", r"macosx-10.9-\1", util.get_platform()
-            )
-
-# MSVS default is dymanic
-if sys.platform == "win32":
-    extra_compile_args.append("/MT")
-
-if "clang" in os.popen("$CC --version 2> /dev/null").read():
-    extra_compile_args.append("-Wno-shorten-64-to-32")
 
 ext_modules = [
     Extension(
@@ -114,7 +70,6 @@ ext_modules = [
         sources=["cyksuid/fast_base62" + suffix, "cyksuid/cbase62.cc"],
         define_macros=ext_macros,
         include_dirs=ext_include_dirs,
-        extra_compile_args=extra_compile_args,
         language="c++",
     ),
     Extension(
@@ -122,21 +77,193 @@ ext_modules = [
         sources=["cyksuid/_ksuid" + suffix],
         define_macros=ext_macros,
         include_dirs=ext_include_dirs,
-        extra_compile_args=extra_compile_args,
         language="c++",
     ),
 ]
 
 
-if USE_CYTHON:
-    ext_cythonize_kwargs["compiler_directives"].update(
-        {
-            "language_level": "3",
-            "embedsignature": True,
-            "binding": True,
-        }
-    )
-    ext_modules = cythonize(ext_modules, **ext_cythonize_kwargs)
+class BuildExt(build_ext.build_ext):
+    def initialize_options(self) -> None:
+        super().initialize_options()
+        self.cython_kwargs: Optional[Dict[str, Any]] = None
+        self.cython_compiler_directives: Optional[dict] = None
+        self.extra_compile_args: List[str] = []
+        self.extra_link_args: List[str] = []
+
+    def finalize_options(self) -> None:
+        super().finalize_options()
+        if self.cython_kwargs is None:
+            self.cython_kwargs = {
+                "force": True,
+            }
+        if self.cython_compiler_directives is None:
+            self.cython_compiler_directives = {
+                "language_level": "3",
+                "embedsignature": True,
+                "binding": True,
+            }
+
+        if IS_DEBUG:
+            self.cython_kwargs.update(
+                {
+                    "gdb_debug": True,
+                    "annotate": True,
+                }
+            )
+            self.cython_compiler_directives.update(
+                {
+                    "linetrace": True,
+                    "profile": True,
+                    "binding": True,
+                }
+            )
+
+    def build_extensions(self) -> None:
+        is_msvc = self.compiler.compiler_type == "msvc"
+        is_clang = hasattr(self.compiler, "compiler_cxx") and (  # noqa: F841
+            "clang++" in self.compiler.compiler_cxx
+        )
+        is_windows = sys.platform[:3] == "win"
+        is_mingw = is_windows and (
+            self.compiler.compiler_type.lower()
+            in ["mingw32", "mingw64", "mingw", "msys", "msys2", "gcc", "g++"]
+        )
+
+        check_args: List[str] = []
+        if sys.platform == "darwin":
+            check_args += [
+                "-Wno-deprecated-register",
+            ]
+
+            # https://developer.apple.com/documentation/xcode_release_notes/xcode_10_release_notes
+            # C++ projects must now migrate to libc++ and are recommended to set a
+            # deployment target of macOS 10.9 or later, or iOS 7 or later.
+            mac_target = str(sysconfig.get_config_var("MACOSX_DEPLOYMENT_TARGET"))
+            if mac_target and (
+                pkg_resources.parse_version(mac_target)
+                < pkg_resources.parse_version("10.9.0")
+            ):
+                os.environ["MACOSX_DEPLOYMENT_TARGET"] = "10.9"
+                os.environ["_PYTHON_HOST_PLATFORM"] = re.sub(
+                    r"macosx-[0-9]+\.[0-9]+-(.+)",
+                    r"macosx-10.9-\1",
+                    util.get_platform(),
+                )
+
+        if is_msvc:
+            self.extra_compile_args += [
+                "/MT",
+                "/O2",
+                "/std:c++14",
+            ]
+        else:
+            check_args += [
+                "-Wno-write-strings",
+                "-Wno-invalid-offsetof",
+                "-Wno-sign-compare",
+                "-Wno-unused-variable",
+                "-Wno-shorten-64-to-32",
+            ]
+            self.add_extra_compile_args(check_args)
+
+        if not is_mingw:
+            self.add_highest_supported_cxx_standard()
+
+        if not IS_DEBUG:
+            self.add_O3()
+            self.add_no_math_errno()
+            self.add_no_trapping_math()
+            if not is_windows:
+                self.add_link_time_optimization()
+
+        for e in self.extensions:
+            e.extra_compile_args += self.extra_compile_args
+
+        if USE_CYTHON:
+            self.extensions = cythonize(
+                self.extensions,
+                compiler_directives=self.cython_compiler_directives,
+                **cast(dict, self.cython_kwargs),
+            )
+        super().build_extensions()
+
+    def add_extra_compile_args(self, args: List[str]) -> None:
+        for arg in args:
+            if self.test_supports_compile_arg(arg):
+                self.extra_compile_args.append(arg)
+
+    def add_highest_supported_cxx_standard(self):
+        cxx17 = "-std=c++17"
+        cxx14 = "-std=gnu++14"
+        cxx11 = "-std=c++11"
+        if self.test_supports_compile_arg(cxx17):
+            self.extra_compile_args.append(cxx17)
+        elif self.test_supports_compile_arg(cxx14):
+            self.extra_compile_args.append(cxx14)
+        elif self.test_supports_compile_arg(cxx11):
+            self.extra_compile_args.append(cxx11)
+        else:
+            msg = "\n\n\nWarning: compiler does not support C++11, compilation of cyksuid might fail without it.\n\n\n"
+            warnings.warn(msg)
+
+    def add_O3(self):
+        O3 = "-O3"
+        if self.test_supports_compile_arg(O3):
+            self.extra_compile_args.append(O3)
+
+    def add_no_math_errno(self):
+        arg_fnme = "-fno-math-errno"
+        if self.test_supports_compile_arg(arg_fnme):
+            self.extra_compile_args.append(arg_fnme)
+            self.extra_link_args.append(arg_fnme)
+
+    def add_no_trapping_math(self):
+        arg_fntm = "-fno-trapping-math"
+        if self.test_supports_compile_arg(arg_fntm):
+            self.extra_compile_args.append(arg_fntm)
+            self.extra_link_args.append(arg_fntm)
+
+    def add_link_time_optimization(self):
+        arg_lto = "-flto"
+        if self.test_supports_compile_arg(arg_lto):
+            self.extra_compile_args.append(arg_lto)
+            self.extra_link_args.append(arg_lto)
+
+    def test_supports_compile_arg(self, comm: Union[str, List[str]]) -> bool:
+        if not hasattr(self.compiler, "compiler_cxx"):
+            return False
+        if not isinstance(comm, list):
+            comm = [comm]
+
+        is_supported = False
+
+        try:
+            if not isinstance(self.compiler.compiler_cxx, list):
+                cmd = list(self.compiler.compiler_cxx)
+            else:
+                cmd = self.compiler.compiler_cxx
+        except Exception as e:
+            print("Error: could not get compiler_cxx: %s" % e)
+            cmd = self.compiler.compiler_cxx
+
+        print("--- Checking compiler support for option '%s'" % " ".join(comm), end="")
+        fname = "cyksuid_compiler_testing.cpp"
+        try:
+            with open(fname, "w") as ftest:
+                ftest.write("int main(int argc, char**argv) {return 0;}\n")
+
+            val_good = subprocess.call(cmd + [fname])
+            val = subprocess.call(cmd + comm + [fname])
+            is_supported = val == val_good
+        except Exception:
+            is_supported = False
+        finally:
+            try:
+                os.remove(fname)
+            except Exception:
+                pass
+        print(" ... %s" % ("yes" if is_supported else "no"))
+        return is_supported
 
 
 setup(
@@ -144,6 +271,7 @@ setup(
     version="2.0.0",
     description="Cython implementation of ksuid",
     ext_modules=ext_modules,
+    cmdclass={"build_ext": BuildExt},
     long_description=open("README.md").read() if os.path.exists("README.md") else "",
     long_description_content_type="text/markdown",
     url="https://github.com/timonwong/cyksuid",
@@ -154,7 +282,7 @@ setup(
     package_data={
         "cyksuid": ["*.pyx", "*.pxd", "*.pyi", "py.typed", "*.h", "*.cc", "*.cpp"],
     },
-    keywords="ksuid",
+    keywords=["ksuid"],
     python_requires=">=3.7",
     classifiers=[
         "Development Status :: 5 - Production/Stable",

